File: ffupdater/src/test/java/de/marmaro/krt/ffupdater/settings/SettingsHelperTest.java
package de.marmaro.krt.ffupdater.settings;

import android.content.Context;
import android.content.SharedPreferences;

import com.github.ivanshafran.sharedpreferencesmock.SPMockBuilder;

import org.junit.Before;
import org.junit.Test;

import java.time.Duration;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;

import de.marmaro.krt.ffupdater.App;
import de.marmaro.krt.ffupdater.device.DeviceEnvironment;
import de.marmaro.krt.ffupdater.utils.Utils;

import static android.os.Build.VERSION_CODES.JELLY_BEAN;
import static android.os.Build.VERSION_CODES.LOLLIPOP;
import static android.os.Build.VERSION_CODES.P;
import static android.os.Build.VERSION_CODES.Q;
import static androidx.appcompat.app.AppCompatDelegate.MODE_NIGHT_AUTO_BATTERY;
import static androidx.appcompat.app.AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM;
import static androidx.appcompat.app.AppCompatDelegate.MODE_NIGHT_NO;
import static androidx.appcompat.app.AppCompatDelegate.MODE_NIGHT_YES;
import static de.marmaro.krt.ffupdater.device.ABI.ARM;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.empty;
import static org.hamcrest.Matchers.is;
import static org.hamcrest.collection.IsIterableContainingInAnyOrder.containsInAnyOrder;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

/**
 * Created by Tobiwan on 03.05.2020.
 */
public class SettingsHelperTest {

    private Context context;
    private SharedPreferences sharedPreferences;

    @Before
    public void setUp() {
        context = mock(Context.class);
        sharedPreferences = new SPMockBuilder().createSharedPreferences();
        when(context.getSharedPreferences(anyString(), anyInt())).thenReturn(sharedPreferences);
    }

    @Test
    public void stringToInt_withNull_returnFallback() {
        assertEquals(42, Utils.stringToInt(null, 42));
    }

    @Test
    public void stringToInt_withEmptyString_returnFallback() {
        assertEquals(24, Utils.stringToInt("", 24));
    }

    @Test
    public void stringToInt_withNumber_returnNumber() {
        assertEquals(30, Utils.stringToInt("30", 42));
    }

    @Test
    public void stringToInt_withNoNumber_returnFallback() {
        assertEquals(12, Utils.stringToInt("hi", 12));
    }

    @Test
    public void isAutomaticCheck_withValueFalse_returnFalse() {
        sharedPreferences.edit().putBoolean("automaticCheck", false).commit();
        assertFalse(new SettingsHelper(context).isAutomaticCheck());
    }

    @Test
    public void isAutomaticCheck_withValueTrue_returnTrue() {
        sharedPreferences.edit().putBoolean("automaticCheck", true).commit();
        assertTrue(new SettingsHelper(context).isAutomaticCheck());
    }

    @Test
    public void getCheckInterval_withNoValue_returnDefaultValue() {
        sharedPreferences.edit().putString("checkInterval", null).commit();
        assertEquals(SettingsHelper.DEFAULT_CHECK_INTERVAL, new SettingsHelper(context).getCheckInterval());

        sharedPreferences.edit().putString("checkInterval", "").commit();
        assertEquals(SettingsHelper.DEFAULT_CHECK_INTERVAL, new SettingsHelper(context).getCheckInterval());
    }

    @Test
    public void getCheckInterval_withInvalidValue_returnDefaultValue() {
        sharedPreferences.edit().putString("checkInterval", "invalid-value").commit();
        assertEquals(SettingsHelper.DEFAULT_CHECK_INTERVAL, new SettingsHelper(context).getCheckInterval());
    }

    @Test
    public void getCheckInterval_withValue_returnDefaultValue() {
        sharedPreferences.edit().putString("checkInterval", "42").commit();
        assertEquals(Duration.ofMinutes(42), new SettingsHelper(context).getCheckInterval());
    }

    @Test
    public void getDisableApps_withNoApps_returnEmptySet() {
        sharedPreferences.edit().putStringSet("disabledApps", null).commit();
        assertTrue(new SettingsHelper(context).getDisableApps().isEmpty());

        sharedPreferences.edit().putStringSet("disabledApps", new HashSet<>()).commit();
        assertTrue(new SettingsHelper(context).getDisableApps().isEmpty());
    }

    @Test
    public void getDisableApps_withSomeApps_returnApps() {
        sharedPreferences.edit().putStringSet("disableApps",
                new HashSet<>(Arrays.asList("FIREFOX_KLAR", "FIREFOX_BETA", "LOCKWISE"))).commit();
        assertThat(new SettingsHelper(context).getDisableApps(), containsInAnyOrder(App.FIREFOX_KLAR, App.FIREFOX_BETA, App.LOCKWISE));
    }

    @Test
    public void getDisableApps_withInvalidApps_ignoreThem() {
        sharedPreferences.edit().putStringSet("disableApps", Utils.createSet("UNKNOWN_APP")).commit();
        assertThat(new SettingsHelper(context).getDisableApps(), is(empty()));
    }

    @Test
    public void getDisableApps_withAllApps_returnApps() {
        sharedPreferences.edit().putStringSet("disableApps", new HashSet<>(Arrays.asList(
                "FIREFOX_KLAR", "FIREFOX_FOCUS", "FIREFOX_LITE", "FIREFOX_RELEASE", "FIREFOX_BETA", "FIREFOX_NIGHTLY",
                "LOCKWISE", "BRAVE"
        ))).commit();
        assertThat(new SettingsHelper(context).getDisableApps(), containsInAnyOrder(App.values()));
    }

    @Test
    public void getThemePreference_withNoValue_BelowJellyBean_returnDefaultValue() {
        final DeviceEnvironment environment = mock(DeviceEnvironment.class);

        sharedPreferences.edit().putString("themePreference", null).commit();
        assertEquals(MODE_NIGHT_NO, new SettingsHelper(context).getThemePreference(environment));

        sharedPreferences.edit().putString("themePreference", "").commit();
        assertEquals(MODE_NIGHT_NO, new SettingsHelper(context).getThemePreference(environment));
    }

    @Test
    public void getThemePreference_withNoValue_JellyBean_returnDefaultValue() {
        final DeviceEnvironment environment = new DeviceEnvironment(Collections.singletonList(ARM), JELLY_BEAN);

        sharedPreferences.edit().putString("themePreference", null).commit();
        assertEquals(MODE_NIGHT_NO, new SettingsHelper(context).getThemePreference(environment));

        sharedPreferences.edit().putString("themePreference", "").commit();
        assertEquals(MODE_NIGHT_NO, new SettingsHelper(context).getThemePreference(environment));
    }

    @Test
    public void getThemePreference_withNoValue_Lollipop_returnDefaultValue() {
        final DeviceEnvironment environment = new DeviceEnvironment(Collections.singletonList(ARM), LOLLIPOP);

        sharedPreferences.edit().putString("themePreference", null).commit();
        assertEquals(MODE_NIGHT_AUTO_BATTERY, new SettingsHelper(context).getThemePreference(environment));

        sharedPreferences.edit().putString("themePreference", "").commit();
        assertEquals(MODE_NIGHT_AUTO_BATTERY, new SettingsHelper(context).getThemePreference(environment));
    }

    @Test
    public void getThemePreference_withNoValue_AndroidP_returnDefaultValue() {
        final DeviceEnvironment environment = new DeviceEnvironment(Collections.singletonList(ARM), P);

        sharedPreferences.edit().putString("themePreference", null).commit();
        assertEquals(MODE_NIGHT_AUTO_BATTERY, new SettingsHelper(context).getThemePreference(environment));

        sharedPreferences.edit().putString("themePreference", "").commit();
        assertEquals(MODE_NIGHT_AUTO_BATTERY, new SettingsHelper(context).getThemePreference(environment));
    }

    @Test
    public void getThemePreference_withNoValue_AndroidQ_returnDefaultValue() {
        final DeviceEnvironment environment = new DeviceEnvironment(Collections.singletonList(ARM), Q);

        sharedPreferences.edit().putString("themePreference", null).commit();
        assertEquals(MODE_NIGHT_FOLLOW_SYSTEM, new SettingsHelper(context).getThemePreference(environment));

        sharedPreferences.edit().putString("themePreference", "").commit();
        assertEquals(MODE_NIGHT_FOLLOW_SYSTEM, new SettingsHelper(context).getThemePreference(environment));
    }

    @Test
    public void getThemePreference_withValue_returnValue() {
        final DeviceEnvironment deviceEnvironment = new DeviceEnvironment(Collections.singletonList(ARM), 30);
        sharedPreferences.edit().putString("themePreference", String.valueOf(MODE_NIGHT_YES)).commit();
        assertEquals(MODE_NIGHT_YES, new SettingsHelper(context).getThemePreference(deviceEnvironment));
    }

    @Test
    public void getThemePreference_withIncorrectValue_returnDefaultValueForApi30() {
        final DeviceEnvironment deviceEnvironment = new DeviceEnvironment(Collections.singletonList(ARM), 30);
        sharedPreferences.edit().putString("themePreference", "exception").commit();
        assertEquals(MODE_NIGHT_FOLLOW_SYSTEM, new SettingsHelper(context).getThemePreference(deviceEnvironment));
    }
}

File: ffupdater/src/test/java/de/marmaro/krt/ffupdater/settings/SettingsHelperTest.kt
package de.marmaro.krt.ffupdater.settings

import android.content.Context
import android.content.SharedPreferences
import androidx.appcompat.app.AppCompatDelegate.*
import com.github.ivanshafran.sharedpreferencesmock.SPMockBuilder
import de.marmaro.krt.ffupdater.app.App
import de.marmaro.krt.ffupdater.device.DeviceSdkTester
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.mockkObject
import io.mockk.unmockkObject
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import java.time.Duration
import java.util.stream.Stream

@ExtendWith(MockKExtension::class)
class SettingsHelperTest {

    @MockK
    lateinit var context: Context
    private lateinit var sharedPreferences: SharedPreferences

    @BeforeEach
    fun setUp() {
        sharedPreferences = SPMockBuilder().createSharedPreferences()
        every { context.getSharedPreferences(any(), any()) } returns sharedPreferences
        every { context.packageName } returns "de.marmaro.krt.ffupdater"
    }

    companion object {
        @JvmStatic
        @BeforeAll
        internal fun beforeAll() {
            mockkObject(DeviceSdkTester)
        }

        @JvmStatic
        @AfterAll
        internal fun afterAll() {
            unmockkObject(DeviceSdkTester)
        }

        @JvmStatic
        fun excludedAppsFromBackgroundUpdateCheck_data(): Stream<Arguments> = Stream.of(
            Arguments.of(App.BRAVE, "BRAVE"),
            Arguments.of(App.BRAVE_BETA, "BRAVE_BETA"),
            Arguments.of(App.BRAVE_NIGHTLY, "BRAVE_NIGHTLY"),
            Arguments.of(App.BROMITE, "BROMITE"),
            Arguments.of(App.FFUPDATER, "FFUPDATER"),
            Arguments.of(App.FIREFOX_BETA, "FIREFOX_BETA"),
            Arguments.of(App.FIREFOX_FOCUS, "FIREFOX_FOCUS"),
            Arguments.of(App.FIREFOX_KLAR, "FIREFOX_KLAR"),
            Arguments.of(App.FIREFOX_NIGHTLY, "FIREFOX_NIGHTLY"),
            Arguments.of(App.FIREFOX_RELEASE, "FIREFOX_RELEASE"),
            Arguments.of(App.ICERAVEN, "ICERAVEN"),
            Arguments.of(App.LOCKWISE, "LOCKWISE"),
            Arguments.of(App.UNGOOGLED_CHROMIUM, "UNGOOGLED_CHROMIUM"),
            Arguments.of(App.VIVALDI, "VIVALDI"),
        )
    }

    @Test
    fun `isForegroundUpdateCheckOnMeteredAllowed test default value`() {
        assertTrue(SettingsHelper(context).isForegroundUpdateCheckOnMeteredAllowed)
    }

    @Test
    fun `isForegroundUpdateCheckOnMeteredAllowed return true`() {
        sharedPreferences.edit().putBoolean("foreground__update_check__metered", true).commit()
        assertTrue(SettingsHelper(context).isForegroundUpdateCheckOnMeteredAllowed)
    }

    @Test
    fun `isForegroundUpdateCheckOnMeteredAllowed return false`() {
        sharedPreferences.edit().putBoolean("foreground__update_check__metered", false).commit()
        assertFalse(SettingsHelper(context).isForegroundUpdateCheckOnMeteredAllowed)
    }

    @Test
    fun `isForegroundUpdateCheckOnMeteredAllowed dont cache value`() {
        val settingsHelper = SettingsHelper(context)

        sharedPreferences.edit().putBoolean("foreground__update_check__metered", false).commit()
        assertFalse(settingsHelper.isForegroundUpdateCheckOnMeteredAllowed)

        sharedPreferences.edit().putBoolean("foreground__update_check__metered", true).commit()
        assertTrue(settingsHelper.isForegroundUpdateCheckOnMeteredAllowed)
    }

    @Test
    fun `isForegroundDownloadOnMeteredAllowed test default value`() {
        assertTrue(SettingsHelper(context).isForegroundDownloadOnMeteredAllowed)
    }

    @Test
    fun `isForegroundDownloadOnMeteredAllowed return true`() {
        sharedPreferences.edit().putBoolean("foreground__download__metered", true).commit()
        assertTrue(SettingsHelper(context).isForegroundDownloadOnMeteredAllowed)
    }

    @Test
    fun `isForegroundDownloadOnMeteredAllowed return false`() {
        sharedPreferences.edit().putBoolean("foreground__download__metered", false).commit()
        assertFalse(SettingsHelper(context).isForegroundDownloadOnMeteredAllowed)
    }

    @Test
    fun `isForegroundDownloadOnMeteredAllowed dont cache value`() {
        val settingsHelper = SettingsHelper(context)

        sharedPreferences.edit().putBoolean("foreground__download__metered", false).commit()
        assertFalse(settingsHelper.isForegroundDownloadOnMeteredAllowed)

        sharedPreferences.edit().putBoolean("foreground__download__metered", true).commit()
        assertTrue(settingsHelper.isForegroundDownloadOnMeteredAllowed)
    }

    @Test
    fun `isBackgroundUpdateCheckEnabled test default value`() {
        assertTrue(SettingsHelper(context).isBackgroundUpdateCheckEnabled)
    }

    @Test
    fun `isBackgroundUpdateCheckEnabled return true`() {
        sharedPreferences.edit().putBoolean("background__update_check__enabled", true).commit()
        assertTrue(SettingsHelper(context).isBackgroundUpdateCheckEnabled)
    }

    @Test
    fun `isBackgroundUpdateCheckEnabled return false`() {
        sharedPreferences.edit().putBoolean("background__update_check__enabled", false).commit()
        assertFalse(SettingsHelper(context).isBackgroundUpdateCheckEnabled)
    }

    @Test
    fun `isBackgroundUpdateCheckEnabled dont cache value`() {
        val settingsHelper = SettingsHelper(context)

        sharedPreferences.edit().putBoolean("background__update_check__enabled", false).commit()
        assertFalse(settingsHelper.isBackgroundUpdateCheckEnabled)

        sharedPreferences.edit().putBoolean("background__update_check__enabled", true).commit()
        assertTrue(settingsHelper.isBackgroundUpdateCheckEnabled)
    }

    @Test
    fun `backgroundUpdateCheckInterval with no settings`() {
        assertEquals(Duration.ofHours(6), SettingsHelper(context).backgroundUpdateCheckInterval)
    }

    @Test
    fun `backgroundUpdateCheckInterval with invalid setting null`() {
        sharedPreferences.edit().putString("background__update_check__interval", null).commit()
        assertEquals(Duration.ofHours(6), SettingsHelper(context).backgroundUpdateCheckInterval)
    }

    @Test
    fun `backgroundUpdateCheckInterval with invalid setting empty string`() {
        sharedPreferences.edit().putString("background__update_check__interval", "").commit()
        assertEquals(Duration.ofHours(6), SettingsHelper(context).backgroundUpdateCheckInterval)
    }

    @Test
    fun `backgroundUpdateCheckInterval with invalid setting string`() {
        sharedPreferences.edit().putString("background__update_check__interval", "lorem ipsum").commit()
        assertEquals(Duration.ofHours(6), SettingsHelper(context).backgroundUpdateCheckInterval)
    }

    @Test
    fun `backgroundUpdateCheckInterval with 42 minutes`() {
        sharedPreferences.edit().putString("background__update_check__interval", "42").commit()
        assertEquals(Duration.ofMinutes(42), SettingsHelper(context).backgroundUpdateCheckInterval)
    }

    @Test
    fun `backgroundUpdateCheckInterval with too low value`() {
        sharedPreferences.edit().putString("background__update_check__interval", "-1").commit()
        assertEquals(Duration.ofMinutes(15), SettingsHelper(context).backgroundUpdateCheckInterval)
    }

    @Test
    fun `backgroundUpdateCheckInterval with too high value`() {
        sharedPreferences.edit().putString("background__update_check__interval", "100000").commit()
        assertEquals(Duration.ofDays(28), SettingsHelper(context).backgroundUpdateCheckInterval)
    }

    @Test
    internal fun `isBackgroundUpdateCheckOnMeteredAllowed default value`() {
        assertTrue(SettingsHelper(context).isBackgroundUpdateCheckOnMeteredAllowed)
    }

    @Test
    internal fun `isBackgroundUpdateCheckOnMeteredAllowed with false`() {
        sharedPreferences.edit().putBoolean("background__update_check__metered", false).commit()
        assertFalse(SettingsHelper(context).isBackgroundUpdateCheckOnMeteredAllowed)
    }

    @Test
    internal fun `isBackgroundUpdateCheckOnMeteredAllowed with true`() {
        sharedPreferences.edit().putBoolean("background__update_check__metered", true).commit()
        assertTrue(SettingsHelper(context).isBackgroundUpdateCheckOnMeteredAllowed)
    }

    @Test
    fun `isBackgroundUpdateCheckOnMeteredAllowed with changing value`() {
        val settingsHelper = SettingsHelper(context)
        sharedPreferences.edit().putBoolean("background__update_check__metered", false).commit()
        assertFalse(settingsHelper.isBackgroundUpdateCheckOnMeteredAllowed)
        sharedPreferences.edit().putBoolean("background__update_check__metered", true).commit()
        assertTrue(settingsHelper.isBackgroundUpdateCheckOnMeteredAllowed)
    }

    @Test
    fun `isBackgroundDownloadEnabled default value`() {
        assertTrue(SettingsHelper(context).isBackgroundDownloadEnabled)
    }

    @Test
    fun `isBackgroundDownloadEnabled with true`() {
        sharedPreferences.edit().putBoolean("background__download__enabled", true).commit()
        assertTrue(SettingsHelper(context).isBackgroundDownloadEnabled)
    }

    @Test
    fun `isBackgroundDownloadEnabled with false`() {
        sharedPreferences.edit().putBoolean("background__download__enabled", false).commit()
        assertFalse(SettingsHelper(context).isBackgroundDownloadEnabled)
    }

    @Test
    fun `isBackgroundDownloadEnabled with changing values`() {
        val settingsHelper = SettingsHelper(context)
        sharedPreferences.edit().putBoolean("background__download__enabled", false).commit()
        assertFalse(settingsHelper.isBackgroundDownloadEnabled)
        sharedPreferences.edit().putBoolean("background__download__enabled", true).commit()
        assertTrue(settingsHelper.isBackgroundDownloadEnabled)
    }

    @Test
    fun `isBackgroundDownloadOnMeteredAllowed default value`() {
        assertFalse(SettingsHelper(context).isBackgroundDownloadOnMeteredAllowed)
    }

    @Test
    fun `isBackgroundDownloadOnMeteredAllowed with true`() {
        sharedPreferences.edit().putBoolean("background__download__metered", true).commit()
        assertTrue(SettingsHelper(context).isBackgroundDownloadOnMeteredAllowed)
    }

    @Test
    fun `isBackgroundDownloadOnMeteredAllowed with false`() {
        sharedPreferences.edit().putBoolean("background__download__metered", false).commit()
        assertFalse(SettingsHelper(context).isBackgroundDownloadOnMeteredAllowed)
    }

    @Test
    fun `isBackgroundDownloadOnMeteredAllowed with changing values`() {
        val settingsHelper = SettingsHelper(context)
        sharedPreferences.edit().putBoolean("background__download__metered", false).commit()
        assertFalse(settingsHelper.isBackgroundDownloadOnMeteredAllowed)
        sharedPreferences.edit().putBoolean("background__download__metered", true).commit()
        assertTrue(settingsHelper.isBackgroundDownloadOnMeteredAllowed)
    }

    @Test
    fun `isBackgroundInstallEnabled default value`() {
        assertFalse(SettingsHelper(context).isBackgroundInstallationEnabled)
    }

    @Test
    fun `isBackgroundInstallEnabled with true`() {
        sharedPreferences.edit().putBoolean("background__installation__enabled", true).commit()
        assertTrue(SettingsHelper(context).isBackgroundInstallationEnabled)
    }

    @Test
    fun `isBackgroundInstallEnabled with false`() {
        sharedPreferences.edit().putBoolean("background__installation__enabled", false).commit()
        assertFalse(SettingsHelper(context).isBackgroundInstallationEnabled)
    }

    @Test
    fun `isBackgroundInstallEnabled with changing values`() {
        val settingsHelper = SettingsHelper(context)
        sharedPreferences.edit().putBoolean("background__installation__enabled", false).commit()
        assertFalse(settingsHelper.isBackgroundInstallationEnabled)
        sharedPreferences.edit().putBoolean("background__installation__enabled", true).commit()
        assertTrue(settingsHelper.isBackgroundInstallationEnabled)
    }

    @Test
    fun `excludedAppsFromBackgroundUpdateCheck with default value`() {
        assertTrue(SettingsHelper(context).excludedAppsFromBackgroundUpdateCheck.isEmpty())
    }

    @Test
    fun `excludedAppsFromBackgroundUpdateCheck with null`() {
        sharedPreferences.edit().putStringSet("background__update_check__excluded_apps", null).commit()
        assertTrue(SettingsHelper(context).excludedAppsFromBackgroundUpdateCheck.isEmpty())
    }

    @Test
    fun `excludedAppsFromBackgroundUpdateCheck with empty set`() {
        sharedPreferences.edit().putStringSet("background__update_check__excluded_apps", setOf()).commit()
        assertTrue(SettingsHelper(context).excludedAppsFromBackgroundUpdateCheck.isEmpty())
    }

    @ParameterizedTest(name = "excludedAppsFromBackgroundUpdateCheck with app \"{0}\"")
    @MethodSource("excludedAppsFromBackgroundUpdateCheck_data")
    fun `excludedAppsFromBackgroundUpdateCheck with app X`(app: App, name: String) {
        sharedPreferences.edit()
            .putStringSet("background__update_check__excluded_apps", setOf(name))
            .commit()
        val disabledApps = SettingsHelper(context).excludedAppsFromBackgroundUpdateCheck
        assertTrue(app in disabledApps)
    }

    @Test
    fun `excludedAppsFromBackgroundUpdateCheck with invalid value`() {
        sharedPreferences.edit().putStringSet("background__update_check__excluded_apps", setOf("invalid"))
            .commit()
        val disabledApps = SettingsHelper(context).excludedAppsFromBackgroundUpdateCheck
        assertTrue(disabledApps.isEmpty())
    }

    @Test
    fun `excludedAppsFromBackgroundUpdateCheck with all apps`() {
        sharedPreferences.edit().putStringSet(
            "background__update_check__excluded_apps",
            setOf(
                "BRAVE",
                "BRAVE_BETA",
                "BRAVE_NIGHTLY",
                "BROMITE",
                "FFUPDATER",
                "FIREFOX_BETA",
                "FIREFOX_FOCUS",
                "FIREFOX_KLAR",
                "FIREFOX_NIGHTLY",
                "FIREFOX_RELEASE",
                "ICERAVEN",
                "LOCKWISE",
                "UNGOOGLED_CHROMIUM",
                "VIVALDI",
            )
        ).commit()
        assertEquals(
            App.values().toList().sorted(),
            SettingsHelper(context).excludedAppsFromBackgroundUpdateCheck.sorted()
        )
    }

    @Test
    fun getThemePreference_userHasNotChangedSetting_AndroidPAndBelow_returnDefaultValue() {
        every { DeviceSdkTester.supportsAndroid10() } returns false
        assertEquals(MODE_NIGHT_AUTO_BATTERY, SettingsHelper(context).themePreference)
    }

    @Test
    fun getThemePreference_userHasNotChangedSetting_AndroidQAndHigher_returnDefaultValue() {
        every { DeviceSdkTester.supportsAndroid10() } returns true
        assertEquals(MODE_NIGHT_FOLLOW_SYSTEM, SettingsHelper(context).themePreference)
    }

    @Test
    fun getThemePreference_withInvalidValue_null_returnDefault() {
        sharedPreferences.edit().putString("foreground__theme_preference", null).commit()

        every { DeviceSdkTester.supportsAndroid10() } returns false
        assertEquals(MODE_NIGHT_AUTO_BATTERY, SettingsHelper(context).themePreference)

        every { DeviceSdkTester.supportsAndroid10() } returns true
        assertEquals(MODE_NIGHT_FOLLOW_SYSTEM, SettingsHelper(context).themePreference)
    }

    @Test
    fun getThemePreference_withInvalidValue_emptyString_returnDefault() {
        sharedPreferences.edit().putString("foreground__theme_preference", "").commit()

        every { DeviceSdkTester.supportsAndroid10() } returns false
        assertEquals(MODE_NIGHT_AUTO_BATTERY, SettingsHelper(context).themePreference)

        every { DeviceSdkTester.supportsAndroid10() } returns true
        assertEquals(MODE_NIGHT_FOLLOW_SYSTEM, SettingsHelper(context).themePreference)
    }

    @Test
    fun getThemePreference_withInvalidValue_text_returnDefault() {
        sharedPreferences.edit().putString("foreground__theme_preference", "lorem ipsum").commit()

        every { DeviceSdkTester.supportsAndroid10() } returns false
        assertEquals(MODE_NIGHT_AUTO_BATTERY, SettingsHelper(context).themePreference)

        every { DeviceSdkTester.supportsAndroid10() } returns true
        assertEquals(MODE_NIGHT_FOLLOW_SYSTEM, SettingsHelper(context).themePreference)
    }

    @Test
    fun getThemePreference_withInvalidValue_nonExistingNumber_returnDefault() {
        sharedPreferences.edit().putString("foreground__theme_preference", "6").commit()

        every { DeviceSdkTester.supportsAndroid10() } returns false
        assertEquals(MODE_NIGHT_AUTO_BATTERY, SettingsHelper(context).themePreference)

        every { DeviceSdkTester.supportsAndroid10() } returns true
        assertEquals(MODE_NIGHT_FOLLOW_SYSTEM, SettingsHelper(context).themePreference)
    }

    @Test
    fun getThemePreference_withValidValue_MODE_NIGHT_FOLLOW_SYSTEM_returnValue() {
        sharedPreferences.edit().putString("foreground__theme_preference", "-1").commit()

        every { DeviceSdkTester.supportsAndroid10() } returns false
        assertEquals(MODE_NIGHT_FOLLOW_SYSTEM, SettingsHelper(context).themePreference)

        every { DeviceSdkTester.supportsAndroid10() } returns true
        assertEquals(MODE_NIGHT_FOLLOW_SYSTEM, SettingsHelper(context).themePreference)
    }

    @Test
    fun getThemePreference_withValidValue_MODE_NIGHT_NO_returnValue() {
        sharedPreferences.edit().putString("foreground__theme_preference", "1").commit()

        every { DeviceSdkTester.supportsAndroid10() } returns false
        assertEquals(MODE_NIGHT_NO, SettingsHelper(context).themePreference)

        every { DeviceSdkTester.supportsAndroid10() } returns true
        assertEquals(MODE_NIGHT_NO, SettingsHelper(context).themePreference)
    }

    @Test
    fun getThemePreference_withValidValue_MODE_NIGHT_YES_returnValue() {
        sharedPreferences.edit().putString("foreground__theme_preference", "2").commit()

        every { DeviceSdkTester.supportsAndroid10() } returns false
        assertEquals(MODE_NIGHT_YES, SettingsHelper(context).themePreference)

        every { DeviceSdkTester.supportsAndroid10() } returns true
        assertEquals(MODE_NIGHT_YES, SettingsHelper(context).themePreference)
    }

    @Test
    fun getThemePreference_withValidValue_MODE_NIGHT_AUTO_BATTERY_returnValue() {
        sharedPreferences.edit().putString("foreground__theme_preference", "3").commit()

        every { DeviceSdkTester.supportsAndroid10() } returns false
        assertEquals(MODE_NIGHT_AUTO_BATTERY, SettingsHelper(context).themePreference)

        every { DeviceSdkTester.supportsAndroid10() } returns true
        assertEquals(MODE_NIGHT_AUTO_BATTERY, SettingsHelper(context).themePreference)
    }
}